include/manif/impl/sgal3/SGal3Tangent_base.h

```diff
@@ -0,0 +1,635 @@
+  const ConstLinBlock lin() const;
+
+  //! @brief Get the angular part.
+  AngBlock ang();
+  const ConstAngBlock ang() const;
+
+  //! @brief Get the linear acceleration part
+  LinBlock lin2();
+  const ConstLinBlock lin2() const;
+
+  Scalar t() const;
+
+public: /// @todo make protected
+
+  const Eigen::Map<const SO3Tangent<Scalar>> asSO3() const {
+    return Eigen::Map<const SO3Tangent<Scalar>>(coeffs().data()+6);
+  }
+
+  Eigen::Map<SO3Tangent<Scalar>> asSO3() {
+    return Eigen::Map<SO3Tangent<Scalar>>(coeffs().data()+6);
+  }
+
+  static void fillE(
+    Eigen::Ref<Eigen::Matrix<Scalar, 3, 3>> E,
+    const Eigen::Map<const SO3Tangent<Scalar>>& so3
+  );
+
+  // static void fillL(
+  //   Eigen::Ref<Eigen::Matrix<Scalar, 3, 3>> E,
+  //   const Eigen::Map<const SO3Tangent<Scalar>>& so3
+  // );
+
+  // static void fillS(
+    // Eigen::Ref<Eigen::Matrix<Scalar, 3, 3>> S,
+    // const Eigen::Map<const SO3Tangent<Scalar>>& so3,
+    // linearVel,
+    // t
+  // );
+};
+
+template <typename _Derived>
+typename SGal3TangentBase<_Derived>::LieGroup
+SGal3TangentBase<_Derived>::exp(OptJacobianRef J_m_t) const {
+  if (J_m_t) {
+    *J_m_t = rjac();
+  }
+
+  const Eigen::Map<const SO3Tangent<Scalar>> so3 = asSO3();
+  const typename SO3<Scalar>::Jacobian so3_ljac = so3.ljac();
+
+  Eigen::Matrix<Scalar, 3, 3> E;
+  fillE(E, so3);
+
+  return LieGroup(
+    so3_ljac * lin() + (E * (t() * lin2())),
+    so3.exp(),
+    so3_ljac * lin2(),
+    t()
+  );
+}
+
+template <typename _Derived>
+typename SGal3TangentBase<_Derived>::LieGroup
+SGal3TangentBase<_Derived>::retract(OptJacobianRef J_m_t) const {
+  return exp(J_m_t);
+}
+
+template <typename _Derived>
+typename SGal3TangentBase<_Derived>::LieAlg
+SGal3TangentBase<_Derived>::hat() const {
+  LieAlg sgal3;
+
+  sgal3.template topLeftCorner<3, 3>() = skew(ang());
+  sgal3.template block<3, 1>(0, 3) = lin2();
+  sgal3.template topRightCorner<3, 1>() = lin();
+  sgal3(3, 4) = t();
+
+  sgal3.template bottomLeftCorner<2, 4>().setZero();
+  sgal3(4, 4) = Scalar(0);
+
+  return sgal3;
+}
+
+template <typename _Derived>
+typename SGal3TangentBase<_Derived>::Jacobian
+SGal3TangentBase<_Derived>::rjac() const {
+
+  return (-*this).ljac();
+
+  // Those were verified against auto diff
+
+  // Jacobian Jr = Jacobian::Zero();
+  // Jr.template topLeftCorner<3, 3>() = asSO3().rjac();
+  // // Jr.template block<3, 3>(0, 3) = ??;
+  // // Jr.template block<3, 3>(0, 6) = ??;
+
+  // // Jr.template block<3, 1>(0, 9) = ??;
+
+  // Jr.template block<3, 3>(3, 3) = Jr.template topLeftCorner<3,3>();
+  // SE3Tangent<Scalar>::fillQ(
+  //   Jr.template block<3, 3>(3, 6), -coeffs().template segment<6>(3)
+  // );
+
+  // Jr.template block<3, 3>(6, 6) = Jr.template topLeftCorner<3,3>();
+
+  // Jr(9, 9) = Scalar(1);
+
+  // Jr.template bottomLeftCorner<7, 3>().setZero();
+  // Jr.template block<4, 3>(6, 3).setZero();
+  // Jr.template block<1, 3>(9, 6).setZero();
+  // Jr.template block<6, 1>(3, 9).setZero();
```

Comment:
Do we want all these coments?

Reply:
Yeah I'll keep those as they are the actual closed-form of the right Jac. Now there are commented because of course it's missing some blocks.